vhost-device-rng/src/main.rs

```diff
@@ -57,58 +58,82 @@ struct RngArgs {
     socket_count: u32,
 
     // Location of vhost-user Unix domain socket. This is suffixed by 0,1,2..socket_count-1.
-    #[clap(short, long)]
-    socket_path: String,
+    #[clap(short, long, value_name = "SOCKET")]
+    socket_path: PathBuf,
 
     // Where to get the RNG data from. Defaults to /dev/urandom.
     #[clap(short = 'f', long, default_value = "/dev/urandom")]
-    rng_source: String,
+    rng_source: PathBuf,
 }
 
 #[derive(Clone, Debug, Eq, PartialEq)]
 pub(crate) struct VuRngConfig {
     pub period_ms: u128,
     pub max_bytes: usize,
     pub count: u32,
-    pub socket_path: String,
-    pub rng_source: String,
+    pub socket_path: PathBuf,
+    pub rng_source: PathBuf,
 }
 
 impl TryFrom<RngArgs> for VuRngConfig {
     type Error = Error;
 
     fn try_from(args: RngArgs) -> Result<Self> {
-        if args.period == 0 || args.period > VHU_RNG_MAX_PERIOD_MS {
-            return Err(Error::InvalidPeriodInput(args.period));
+        let RngArgs {
+            period,
+            max_bytes,
+            socket_count,
+            socket_path,
+            rng_source,
+        } = args;
+
+        if period == 0 || period > VHU_RNG_MAX_PERIOD_MS {
+            return Err(Error::InvalidPeriodInput(period));
         }
 
-        if args.socket_count == 0 {
-            return Err(Error::InvalidSocketCount(args.socket_count));
+        if socket_count == 0 {
+            return Err(Error::InvalidSocketCount(socket_count));
         }
 
         // Divide available bandwidth by the number of threads in order
         // to avoid overwhelming the HW.
-        let max_bytes = args.max_bytes / args.socket_count as usize;
-        let socket_path = args.socket_path.trim().to_string();
-        let rng_source = args.rng_source.trim().to_string();
+        let max_bytes = max_bytes / socket_count as usize;
 
         Ok(VuRngConfig {
-            period_ms: args.period,
+            period_ms: period,
             max_bytes,
-            count: args.socket_count,
+            count: socket_count,
             socket_path,
             rng_source,
         })
     }
 }
 
+impl VuRngConfig {
+    pub fn generate_socket_paths(&self) -> Vec<PathBuf> {
+        let socket_file_name = self
+            .socket_path
+            .file_name()
+            .expect("socket_path has no filename.");
+        let socket_file_parent = self
+            .socket_path
+            .parent()
+            .expect("socket_path has no parent directory.");
+
+        let make_socket_path = |i: u32| -> PathBuf {
+            let mut file_name = socket_file_name.to_os_string();
+            file_name.push(std::ffi::OsStr::new(&i.to_string()));
```

Comment:
OsStr::from(i.to_string()) probably does the conversion in "zero-copy"

Reply:
OsStr is a borrowed ref, so there's no copying. Do you mean something else?

Also OsStr doesn't implement From<String> [from what I see here](https://doc.rust-lang.org/std/ffi/struct.OsStr.html)...

Reply:
Ah. I read it as OsString... Confusing that str is lower-case but OsStr is camel... Looks good.
Still need to get out of the C++ habit of being afraid of seeing `&` and getting paranoid of whether reference to temporary objects stay alive...